File: compiler.py
import re

TOKEN_TYPES = [
        ['def', r'\bdef\b'],
        ['identifier', r'\b[a-zA-Z]+\b'],
        ['integer', r'\b[0-9]+\b'],
        ['colon', r':'],
        ['oparent', r'\('],
        ['cparent', r'\)'],
        ['comma', r','], 
]

class Tokeinzer:
    def __init__(self, code) -> None:
        self.code = code

    def tokenize(self):
        self.tokens = []
        while self.code:
            self.tokens.append(self.tokenize_one_token())
            self.code = self.code.strip()
        return self.tokens

    def tokenize_one_token(self):
        for t,regex in TOKEN_TYPES:
            p = re.compile(rf'\A({regex})')
            m = p.match(self.code)
            if m:
                value = m.group()
                self.code = self.code[len(value):]
                return Token(t, value)

class Token:
    def __init__(self, type, value) -> None:
        self.type = type
        self.value = value

class Parser:
    def __init__(self, tokens) -> None:
        self.tokens = tokens

    def parse(self):
        return self.parse_def()
    
    def parse_def(self):
        self.consume('def')
        name = self.consume('identifier')
        arg_names = self.parse_arg_names()
        body = self.parse_expr()
        return DefNode(name, arg_names, body)

    def consume(self, expected_type):
        token = self.tokens.pop(0)
        if token.type == expected_type:
            return token.value
        else:
            raise RuntimeError(f"Expected token type {expected_type} but got {token.type}.")
        
    def parse_arg_names(self):
        arg_names = []
        self.consume('oparent')
        if self.peek('identifier'):
            arg_names.append(self.consume('identifier'))
            while self.peek('comma'):
                self.consume('comma')
                arg_names.append(self.consume('identifier'))
        self.consume('cparent')
        self.consume('colon')
        return arg_names

    def parse_expr(self):
        if self.peek('integer'):
            return self.parse_integer()
        if self.peek('identifier') and self.peek('oparent', 1):
            return self.parse_call()
        return self.parse_var_ref()

    def parse_integer(self):
        return IntegerNode(int(self.consume('integer'))).value
    
    def parse_call(self):
        name = self.consume('identifier')
        arg_exprs = self.parse_arg_exprs()
        return CallNode(name, arg_exprs)
    
    def parse_var_ref(self):
        return VarRefNode(self.consume('identifier'))
    
    def parse_arg_exprs(self):
        arg_exprs = []
        self.consume('oparent')
        if self.peek('cparent') != True: 
            arg_exprs.append(self.parse_expr())
            while self.peek('cparent') != True:
                self.consume('comma')
                arg_exprs.append(self.parse_expr())
        self.consume('cparent')
        return arg_exprs

    def peek(self, expected_type, offset=0):
        return self.tokens[offset].type == expected_type

class DefNode:
    def __init__(self, name , arg_names, body) -> None:
        self.name = name
        self.arg_names = arg_names
        self.body = body

class IntegerNode:
    def __init__(self, value) -> None:
        self.value = value

class CallNode:
    def __init__(self, name, arg_exprs) -> None:
        self.name = name
        self.arg_exprs = arg_exprs

class VarRefNode:
    def __init__(self, value) -> None:
        self.value = value

class Generator:
    def __init__(self) -> None:
        pass

    def generate(self, node):
        if isinstance(node, DefNode):
            return "function %s(%s) {return %s};" % (
                node.name, 
                ','.join(node.arg_names), 
                self.generate(node.body))
        
        if isinstance(node, CallNode):
            return "%s(%s)" % (
                node.name, 
                ','.join([self.generate(expr) for expr in node.arg_exprs])
                )
        
        if isinstance(node, int):
            return str(node)

        if isinstance(node, VarRefNode):
            return node.value
        
        raise RuntimeError(f"Unexpected node type: {node.__class__}")


tokens = Tokeinzer(open("test.src").read()).tokenize()
tree = Parser(tokens).parse()
generator = Generator()
generated = generator.generate(tree)
print(generated)

File: compiler.jl
TOKEN_TYPES = [
    (:def, "\\bdef\\b"),
    (:end, "\\bend\\b"),
    (:identifier, "\\b[a-zA-Z]+\\b"),
    (:integer, "\\b[0-9]+\\b"),
    (:oparent, "\\("),
    (:cparent, "\\)"),
    (:comma, ","), 
]

struct Token
    type::Symbol
    value::String
end

struct IntegerNode
    value::Integer
end

struct CallNode
    name::String
    arg_exprs::Vector
end

struct VarRefNode
    value::String
end

struct DefinitionNode
    name::String
    arg_names::Vector
    body::CallNode
end



function tokenize(code::AbstractString)
    tokens = []
    while !isempty(code)
        token, code = tokenize_one_token(code)
        push!(tokens, token)
        code = strip(code)
    end
    return tokens
end

function tokenize_one_token(code::AbstractString)
    for type in TOKEN_TYPES
        pattern = Regex("\\A$(type[2])")
        matched = match(pattern, code)
        if matched !== nothing
            value = matched.match
            code = code[length(value)+1:end]
            return Token(type[1], value), code
        end
    end
end

function parse_def(tokens::Vector)
    consume(tokens, :def)
    name = consume(tokens, :identifier)
    arg_names = parse_arg_names(tokens)
    body = parse_expr(tokens)
    return DefinitionNode(name, arg_names, body)
end

function consume(tokens::Vector, expected_type::Symbol)
    token = splice!(tokens, 1)
    if token.type == expected_type
        return token.value
    else
        error_msg = "Expected token type $expected_type but got $(token.type)."
        throw(error(error_msg))
    end
end

function parse_arg_names(tokens::Vector)
    arg_names = []
    consume(tokens, :oparent)
    if peek(tokens, :identifier)
        push!(arg_names, consume(tokens, :identifier))
        while peek(tokens, :comma)
            consume(tokens, :comma)
            push!(arg_names, consume(tokens, :identifier))
        end
    end
    consume(tokens, :cparent)
    return arg_names
end

function parse_expr(tokens::Vector)
    if peek(tokens, :integer)
        return parse_integer(tokens)
    end
    if peek(tokens, :identifier) & peek(tokens, :oparent, 2)
        return parse_call(tokens)
    end
    return parse_var_ref(tokens)
end

function parse_integer(tokens::Vector)
    return IntegerNode(parse(Int,consume(tokens, :integer)))
end

function parse_call(tokens::Vector)
    name = consume(tokens, :identifier)
    arg_exprs = parse_arg_exprs(tokens)
    return CallNode(name, arg_exprs)
end

function parse_var_ref(tokens::Vector)
    return VarRefNode(consume(tokens,:identifier))
end

function parse_arg_exprs(tokens::Vector)
    arg_exprs = []
    consume(tokens, :oparent)
    if peek(tokens, :cparent) != true
        push!(arg_exprs, parse_expr(tokens))
        while peek(tokens, :cparent) != true
            consume(tokens, :comma)
            push!(arg_exprs, parse_expr(tokens))
        end
    end
    consume(tokens, :cparent)
    return arg_exprs
end

function peek(tokens::Vector, expected_type::Symbol, offset::Integer=1)
    return tokens[offset].type == expected_type
end

function generate(tree)
    if isa(tree, DefinitionNode)
        arg_names = join(tree.arg_names, ",")
        body = generate(tree.body)
        return "def $(tree.name)($(arg_names)):return $(body)"
    end

    if isa(tree, CallNode)
        return "$(tree.name)($(join([generate(expr) for expr in tree.arg_exprs],",")))"
    end

    if isa(tree, IntegerNode)
        return string(tree.value)
    end

    if isa(tree, VarRefNode)
        return tree.value
    end
end

tokens = tokenize(read("test.src", String)) 
tree = parse_def(tokens)
generated = generate(tree)
RUNTIME = "def add(x,y):return x+y"
RUNNING_F = "print(f(1,2))"
pyfile = open("julia_compilation.py", "w")
write(pyfile,RUNTIME * "\n" *generated * "\n" * RUNNING_F) 
close(pyfile)
run(`python3.10 julia_compilation.py`)
